files/edge-core/scripts/edge-core-bootloader.sh

```diff
@@ -85,57 +86,85 @@ function check_snap_refresh() {
 
 echo "Checking for ${UPGRADE_TGZ}"
 if [ -e "${UPGRADE_TGZ}" ]; then
-    echo "Unpacking ${UPGRADE_TGZ}..."
+    if [ -e "${UPGRADE_WORKDIR}" ]; then
+        echo "Deleting old upgrade workdir..."
+        rm -rf "${UPGRADE_WORKDIR}"
+    fi
+    echo "Unpacking ${UPGRADE_TGZ} to ${UPGRADE_WORKDIR}..."
     mkdir -p "${UPGRADE_WORKDIR}"
     tar --no-same-owner -xzf "${UPGRADE_TGZ}" -C "${UPGRADE_WORKDIR}"
     # remove the upgrade tgz file so that we don't fall into an upgrade loop
     rm "${UPGRADE_TGZ}"
-    # TODO: init fota process state tracking
+else
+    echo "No upgrade found...continue to boot"
+    exit 0
 fi
 
 # move into folder and call pre-refresh if exists
-retval=0
-if [ -e "${UPGRADE_WORKDIR}"]; then
+if [ -e "${UPGRADE_WORKDIR}" ]; then
+    echo "Processing upgrade..."
     pushd "${UPGRADE_WORKDIR}"
 
     if [ -x pre-refresh.sh ]; then
         echo "edge-core-bootloader.sh: Running pre-refresh.sh" | systemd-cat -p info -t FOTA-PRE-REFRESH
         ./pre-refresh.sh 2>&1 | systemd-cat -p info -t FOTA-PRE-REFRESH
         retval=$?
-        rm pre-refresh.sh
+        if [ $retval = 0 ]; then
+            rm pre-refresh.sh
+        else
+            popd
+            echo "pre-refresh.sh script failed: ${retval}"
+            rm -rf "${UPGRADE_WORKDIR_FAILED}"
+            mv "${UPGRADE_WORKDIR}" "${UPGRADE_WORKDIR_FAILED}"
+            exit 1
```

Comment:
I'd like to see this entire `if` statement factored out into a function (with parameters: `"pre-refresh"`, `$?`, `<exit_code>`).
The `rm pre-refresh` part can then simply exist below the function call, without an additional `if`, since the function will exit the script on error. Plus, we won't really need a `retval` variable anymore either. Something like this:

```
function check_error() {
    if [ "$2" != 0 ]; then
        popd
        echo "$1 script failed: $2"
        rm -rf "${UPGRADE_WORKDIR_FAILED}"
        mv "${UPGRADE_WORKDIR}" "${UPGRADE_WORKDIR_FAILED}"
        exit $3
    fi
}

...

check_error "pre-refresh" $? 1
rm pre-refresh.sh
```

Reply:
I thought of that too, but the popd wouldn't fit well.  Too much knowledge about the calling context.